atari-DQN/dqn_agent.py

```diff
@@ -0,0 +1,315 @@
+            zero_pad_mat = torch.zeros(self.observation_shape,
+                                       dtype=self.observation_dtype,
+                                       device='cpu')
+            self.history_queue.append(zero_pad_mat)
+
+        # clear per-episode counters
+        self.episode_total_policy_loss = 0.0
+        self.episode_total_optim_steps = 0
+
+        # NOTE: first observation is not stored in memory buffer
+        # TODO: maybe add training step?
+
+    def step(self, observation: np.ndarray) -> int:
+        """
+        The agent takes one step
+
+        :param observation:  initial observation from the environment, should
+                have shape self.observation_shape
+        :return: int denoting action to take at this step
+        """
+        # Cast observation to tensor and add to history
+        observation = torch.tensor(observation, dtype=self.observation_dtype,
+                                   device='cpu')
+        self.history_queue.append(observation)
+
+        self._train_step()
+        # Select action and return int
+        return self._select_action()
+
+    def _select_action(self) -> int:
+        """
+        Select action according to the epsilon greedy policy
+        :return: int denoting action to take
+        """
+
+        # Compute epsilon
+        epsilon = self.epsilon_fn(self.epsilon_decay_period,
+                                  self.training_steps,
+                                  self.min_replay_history,
+                                  self.epsilon_final)
+
+        if random.random() <= epsilon:
+            # random action with probability epsilon
+            return random.randrange(self.num_actions)
+        else:
+            # Construct state
+            state = self._history_queue_to_state()
+            # greedy action
+            action_tensor = self.policy_net(state).max(1)[1].view(1, 1)
+            return action_tensor.item()  # TODO ensure this works
+
+    def _train_step(self):
+        """Runs a single training step.
+        Runs a training op if both:
+          (1) A minimum number of frames have been added to the replay buffer.
+          (2) `training_steps` is a multiple of `update_period`.
+        Also, syncs weights from online to target network if training steps is a
+        multiple of target update period.
+        """
+        # Run a train op at the rate of self.update_period if enough training steps
+        # have been run. This matches the Nature DQN behaviour.
+        if len(self.memory) > self.min_replay_history:
+            if self.training_steps % self.update_period == 0:
+                self._optimize_model()
+                # TODO: Dopamine had a bunch of summary writers here
+            # Update target network
+            if self.training_steps % self.target_update_period == 0:
+                self.target_net.load_state_dict(self.policy_net.state_dict())
+
+        self.training_steps += 1
```

Comment:
Should this be inside of the above "if" statement? That is, `training_steps` should only be incremented if the model actually underwent optimization?